monocle/crawler.py

```diff
@@ -53,31 +58,90 @@ def __init__(self, args, elastic_conn='localhost:9200', elastic_timeout=10):
         elif args.command == 'gerrit_crawler':
             self.repository_el_re = args.repository.lstrip('^')
             self.prf = review.ReviewesFetcher(args.base_url, args.repository)
-        self.setName(self.repository_el_re)
 
     def get_last_updated_date(self):
         change = self.db.get_last_updated(self.repository_el_re)
         if not change:
             return self.updated_since or datetime.now().strftime("%Y-%m-%dT%H:%M:%SZ")
         else:
-            logging.info(
+            log.info(
                 "Most recent change date in the database for %s is %s"
                 % (self.repository_el_re, change['updated_at'])
             )
             return change['updated_at']
 
+    def dump_data(self, data, prefix=None):
+        try:
+            if self.dump_dir:
+                tmpfile = tempfile.NamedTemporaryFile(
+                    dir=self.dump_dir,
+                    prefix=prefix,
+                    suffix='.json',
+                    mode='w',
+                    delete=False,
+                )
+                json.dump(data, tmpfile)
+                tmpfile.close()
+                log.info('data dumped to %s' % tmpfile.name)
+                return tmpfile.name
+        except Exception:
+            log.exception('Unable to dump data')
+        return None
+
     def run_step(self):
         updated_since = self.get_last_updated_date()
-        prs = self.prf.get(updated_since)
-        objects = self.prf.extract_objects(prs)
-        if objects:
-            self.log.info("%s objects will be updated in the database" % len(objects))
-            self.db.update(objects)
+        try:
+            prs = self.prf.get(updated_since)
+        except Exception:
+            log.exception('Unable to get PR data')
+            return
+        if len(prs) > 0:
+            self.dump_data(prs, 'raw_prs_')
+        while len(prs) > 0:
+            try:
+                objects = self.prf.extract_objects(prs)
+                if objects:
+                    log.info(
+                        "%s objects will be updated in the database" % len(objects)
+                    )
+                    self.db.update(objects)
+                break
+            except pullrequest.ExtractPRIssue as err:
+                log.exception('Unable to decode PR')
+                log.error('idx=%d pr=%s' % (err.idx, err.pr))
+                # stop if we were on the last pr
+                if len(prs) == err.idx + 1:
+                    break
+                prs = prs[: err.idx] + prs[err.idx + 1 :]  # noqa
 
```

Comment:
I don't get why we need to dump everything line 99.
Why to process extract_objects prs by bulk. It seems easier to me to skip and dump a failed extract in each gerrit and github crawler code directly. There is already the code to skip in each crawler so we can pass a dump_on_failed function to extract_objects.

Reply:
yes changing that